src/brod_producer.erl

```diff
@@ -56,300 +50,295 @@
+-spec start_link([brod:host()], proplists:proplist(), [term()]) ->
                     {ok, pid()} | {error, any()}.
-start_link(Hosts, RequiredAcks, AckTimeout, ClientId, Debug) ->
-  Args = [Hosts, RequiredAcks, AckTimeout, ClientId, Debug],
-  gen_server:start_link(?MODULE, Args, [{debug, Debug}]).
+start_link(Hosts, Options, Debug) ->
+  case proplists:get_value(producer_id, Options) of
+    undefined ->
+      gen_server:start_link(?MODULE, [Hosts, Options, Debug], [{debug, Debug}]);
+    ProducerId when is_atom(ProducerId) ->
+      gen_server:start_link({local, ProducerId}, ?MODULE, [Hosts, Options, Debug], [{debug, Debug}])
+  end.
 
 -spec stop(pid()) -> ok | {error, any()}.
 stop(Pid) ->
   gen_server:call(Pid, stop, infinity).
 
 -spec produce(pid(), binary(), integer(), [{binary(), binary()}]) ->
-                 {ok, reference()}.
+                 {ok, reference()} | ok.
 produce(Pid, Topic, Partition, KVList) ->
-  Sender = self(),
   Ref = erlang:make_ref(),
-  gen_server:call(Pid, {produce, Sender, Ref, Topic, Partition, KVList},
-                  infinity).
-
--spec debug(pid(), print | string() | none) -> ok.
-%% @doc Enable debugging on producer and its connection to a broker
-%%      debug(Pid, pring) prints debug info on stdout
-%%      debug(Pid, File) prints debug info into a File
-debug(Pid, print) ->
-  do_debug(Pid, {trace, true}),
-  do_debug(Pid, {log, print});
-debug(Pid, File) when is_list(File) ->
-  do_debug(Pid, {trace, true}),
-  do_debug(Pid, {log_to_file, File}).
-
--spec no_debug(pid()) -> ok.
-%% @doc Disable debugging
-no_debug(Pid) ->
-  do_debug(Pid, no_debug).
-
--spec get_sockets(pid()) -> {ok, [#socket{}]}.
-get_sockets(Pid) ->
-  gen_server:call(Pid, get_sockets, infinity).
+  SenderPid = self(),
+  Msg = {produce, Ref, SenderPid, Topic, Partition, KVList},
+  gen_server:call(Pid, Msg, infinity).
+
+subscribe(Pid, WorkerPid, LeaderId) ->
+  gen_server:call(Pid, {subscribe, WorkerPid, LeaderId}, infinity).
+
+get_request(RequestRef) ->
+  [Request] = ets:lookup(?T_REQUESTS, RequestRef),
+  {ok, Request}.
 
 %%%_* gen_server callbacks -----------------------------------------------------
-init([Hosts, RequiredAcks, AckTimeout, ClientId, Debug]) ->
+init([Hosts, Options, Debug]) ->
   erlang:process_flag(trap_exit, true),
-  {ok, #state{ hosts       = Hosts
-             , debug       = Debug
-             , acks        = RequiredAcks
-             , ack_timeout = AckTimeout
-             , client_id   = ensure_binary(ClientId)
+  MaxLeaderQueueLen = proplists:get_value(max_leader_queue_len, Options,
+                                          ?DEFAULT_MAX_LEADER_QUEUE_LEN),
+  TabLeaders = ets:new(t_leaders, [protected]),
+  ets:new(?T_REQUESTS, [protected, named_table, ordered_set,
+                        {keypos, #request.ref}]),
+  {ok, SupPid} = brod_producer_sup:start_link(),
+  {ok, Brokers} = refresh_metadata(Hosts, TabLeaders),
```

Comment:
Do not fetch the metadata or interact with Kafka in any way from the `init` function! If Kafka cannot be reached, this could crash your `init`. The supervisor will not restart a child who's `init` crashed, but terminate instead.

So a single network problem may prevent your entire application, and maybe the entire node from starting up.